src/prepared_discover/prepared_discover_txn.c

```diff
@@ -8,53 +8,90 @@
- *     We have learned that a prepared transaction with a particular ID exists. If this is the first
- *     time it's been noticed, create a transaction corresponding to it. Otherwise return the
- *     matching transaction.
+ * __wt_prepared_discover_find_transaction --
+ *     Find a prepared transaction by its ID in the pending prepared transactions hash map.
  */
 int
-__wt_prepared_discover_find_or_create_transaction(WT_SESSION_IMPL *session,
-  wt_timestamp_t prepare_transaction_id, WT_SESSION_IMPL **prep_sessionp, bool create)
+__wt_prepared_discover_find_transaction(WT_SESSION_IMPL *session, uint64_t prepare_transaction_id,
+  WT_PENDING_PREPARED_ITEM **prepared_item)
 {
     WT_CONNECTION_IMPL *conn;
-    WT_SESSION_IMPL *new_session, *next_session;
+    WT_PENDING_PREPARED_ITEM *item;
+    WT_PENDING_PREPARED_MAP *pending_prepare_items;
     WT_TXN_GLOBAL *txn_global;
-    uint32_t prepared_session_cnt;
-
-    new_session = next_session = NULL;
+    uint64_t bucket;
     conn = S2C(session);
     txn_global = &conn->txn_global;
-
-    if (txn_global->pending_prepared_sessions != NULL) {
-        for (prepared_session_cnt = 0;
-             (next_session = txn_global->pending_prepared_sessions[prepared_session_cnt]) != NULL;
-             prepared_session_cnt++) {
-            if (next_session->txn->prepare_timestamp == prepare_transaction_id) {
-                *prep_sessionp = next_session;
+    pending_prepare_items = &txn_global->pending_prepare_items;
+    if (pending_prepare_items->hash != NULL) {
+        bucket = prepare_transaction_id & (pending_prepare_items->hash_size - 1);
+        TAILQ_FOREACH (item, &pending_prepare_items->hash[bucket], hashq) {
+            if (item->prepared_id == prepare_transaction_id) {
+                *prepared_item = item;
                 return (0);
             }
         }
     }
+    return (WT_NOTFOUND);
+}
+
+/*
+ * __pending_prepare_items_init --
+ *     Initialize pending prepared txn hash map.
+ */
+static int
+__pending_prepare_items_init(
+  WT_SESSION_IMPL *session, WT_PENDING_PREPARED_MAP *pending_prepare_items, u_int hash_size)
+{
+    /* Hash size must be a power of 2 for efficient bucket calculation. */
+    WT_ASSERT(session, (hash_size & (hash_size - 1)) == 0);
+
+    pending_prepare_items->hash_size = hash_size;
+    WT_RET(
+      __wt_calloc_def(session, pending_prepare_items->hash_size, &pending_prepare_items->hash));
+    for (uint64_t i = 0; i < pending_prepare_items->hash_size; i++) {
+        TAILQ_INIT(&pending_prepare_items->hash[i]); /* hash lists */
+    }
+    return (0);
+}
+
+/*
+ * __prepared_discover_find_or_create_transaction --
+ *     We have learned that a prepared transaction with a particular ID exists. If this is the first
+ *     time it's been noticed, create a transaction corresponding to it. Otherwise return the
+ *     matching transaction.
+ */
+static int
+__prepared_discover_find_or_create_transaction(WT_SESSION_IMPL *session,
```

Comment:
I think here we are not creating a transaction. We should find a better name for this and update the comment.

Reply:
renamed to item